crates/chia-datalayer/src/merkle/iterators.rs

```diff
@@ -0,0 +1,547 @@
+            predicate,
+            from_index,
+        }
+    }
+}
+
+impl Iterator for LeftChildFirstIterator<'_> {
+    type Item = Result<(TreeIndex, Block), Error>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // left sibling first, children before parents
+
+        loop {
+            let item = self.stack.pop()?;
+            let block = match try_get_block(self.blob, item.index) {
+                Ok(block) => block,
+                Err(e) => return Some(Err(e)),
+            };
+
+            if let Some(predicate) = self.predicate {
+                if !predicate(&block) {
+                    continue;
+                }
+            }
+
+            match block.node.parent().0 {
+                // TODO: maybe also check the parent considers this the child?
+                Some(index) => {
+                    if item.index != self.from_index && !self.already_queued.contains(&index) {
+                        return Some(Err(Error::ReferenceToUnknownParent()));
+                    }
+                }
+                None => {
+                    if item.index.0 != 0 {
+                        return Some(Err(Error::UnexpectedParentlessNode()));
+                    }
+                }
+            }
+
+            match block.node {
+                Node::Leaf(..) => {
+                    if block.metadata.dirty {
+                        return Some(Err(Error::DirtyLeaf(item.index)));
+                    }
+                    return Some(Ok((item.index, block)));
+                }
+                Node::Internal(ref node) => {
+                    if item.visited {
+                        return Some(Ok((item.index, block)));
+                    };
+
+                    if node.left == node.right
+                        || self.already_queued.contains(&node.left)
+                        || self.already_queued.contains(&node.right)
+                    {
+                        return Some(Err(Error::InvalidChildren()));
+                    }
+
+                    if self.already_queued.contains(&item.index) {
+                        return Some(Err(Error::CycleFound()));
+                    }
+                    self.already_queued.insert(item.index);
+
+                    self.stack.push(LeftChildFirstIteratorItem {
+                        visited: true,
+                        index: item.index,
+                    });
+                    self.stack.push(LeftChildFirstIteratorItem {
+                        visited: false,
+                        index: node.right,
+                    });
+                    self.stack.push(LeftChildFirstIteratorItem {
+                        visited: false,
+                        index: node.left,
+                    });
+                }
+            }
+        }
+    }
+}
+
+pub struct ParentFirstIterator<'a> {
+    blob: &'a [u8],
+    deque: VecDeque<TreeIndex>,
+    already_queued: HashSet<TreeIndex>,
+}
+
+impl<'a> ParentFirstIterator<'a> {
```

Comment:
an idea for a fuzzer:
generate a random tree, iterate it with every kind of iterator, ensure that every node was visited exactly once.